core/modules/user/user.module

```diff
@@ -990,6 +1001,56 @@ function user_account_form_validate($form, &$form_state) {
   }
 }
 
+/**
+ * Validation callback for password constraints.
+ *
+ * @see user_account_form().
+ * @see user_pass_reset_form().
+ */
+function user_password_constraints_validate($form, &$form_state) {
+  if (!array_key_exists('pass', $form_state['values'])) {
+    // User registration forms don't have a password field,
+    // if mail verification is required.
+    return;
+  }
+  $password = trim($form_state['values']['pass']);
+  if ($form['#form_id'] == 'user_profile_form') {
+    if (strlen($password) == 0) {
+      // On profile form and no new password has been set.
+      return;
+    }
+  }
+  $constraints = config_get('system.core', 'pass_constraints');
+  $messages = array();
+  if ($form['#form_id'] == 'user_pass_reset_form') {
+    $account = $form['#account'];
+    $username = $account->name;
+    $email = $account->mail;
+  }
+  else {
+    $username = $form_state['values']['name'];
+    $email = $form_state['values']['mail'];
+  }
+  if (strlen($password) < $constraints['pass_minlength']) {
+    $min = $constraints['pass_minlength'];
+    $messages[] = t('The password is too short, please use at least :min characters.', array(':min' => $min));
+  }
+  if ($constraints['pass_not_username']) {
+    if ($password == $username) {
+      $messages[] = t("Please don't use the username as password.");
```

Comment:
UI text-writing guidelines dictate avoiding the word "please", and using neutral tone where possible. How about something like:

```suggestion
      $messages[] = t("The username cannot be used as password.");
```